haystack/dataclasses/chat_message.py

```diff
@@ -74,6 +99,38 @@ class ToolCallResult:
     origin: ToolCall
     error: bool
 
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Converts ToolCallResult into a dictionary.
+
+        :returns: A dictionary with keys 'result', 'origin', and 'error'.
+        """
+        return {"result": self.result, "origin": self.origin.to_dict(), "error": self.error}
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "ToolCallResult":
+        """
+        Creates a ToolCallResult from a dictionary.
+
+        :param data:
+            The dictionary to build the ToolCallResult object.
+        :returns:
+            The created object.
+        """
+        if not all(x in data for x in ["result", "origin", "error"]):
+            raise ValueError(
+                "Fields `result`, `origin`, `error` are required for ToolCallResult deserialization. "
+                f"Received dictionary with keys {list(data.keys())}"
+            )
+
+        origin = data["origin"]
+        if isinstance(origin, dict):
+            origin = ToolCall.from_dict(origin)
+        elif not (origin is None or isinstance(origin, ToolCall)):
+            raise TypeError("`origin` must be a dict or ToolCall.")
+
+        return ToolCallResult(result=data["result"], origin=origin, error=data["error"])
```

Comment:
To be consistent with our other from_dict methods let's only support origin being in a dict format. It's fine to leave the ValueError check here otherwise we would get a key a access error to `data`
```suggestion
        if not all(x in data for x in ["result", "origin", "error"]):
            raise ValueError(
                "Fields `result`, `origin`, `error` are required for ToolCallResult deserialization. "
                f"Received dictionary with keys {list(data.keys())}"
            )

        return ToolCallResult(result=data["result"], origin=ToolCall.from_dict(data["origin"]), error=data["error"])
```